packages/sdk/client-services/src/packlets/tests/text.test.ts

```diff
@@ -0,0 +1,210 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import * as fc from 'fast-check';
+import { ModelRunSetup } from 'fast-check';
+import waitForExpect from 'wait-for-expect';
+
+import { Client, Document, Text } from '@dxos/client';
+import { PublicKey } from '@dxos/keys';
+import { afterTest, describe, test } from '@dxos/test';
+import { Doc } from '@dxos/text-model';
+import { ComplexMap, ComplexSet, range } from '@dxos/util';
+
+import { joinCommonSpace, TestBuilder } from '../testing';
+
+type Model = {
+  text: string;
+  peers: ComplexSet<PublicKey>;
+};
+
+type Real = {
+  spaceKey: PublicKey;
+  peers: ComplexMap<PublicKey, Client>;
+};
+
+const assertState = async (model: Model, real: Real) => {
+  await waitForExpect(() => {
+    for (const peer of real.peers.values()) {
+      const space = peer.echo.getSpace(real.spaceKey);
+      if (space) {
+        if (!model.peers.has(peer.halo.identity!.identityKey)) {
+          throw new Error(`Expected peer to not be in space: ${peer.halo.identity!.identityKey.truncate()}`);
+        }
+
+        const [document] = space.db.query((obj) => !!obj.content).objects;
+        const text = (document.content.doc as Doc).getText('utf8');
+        if (text.toString() !== model.text) {
+          throw new Error(`Expected text to be "${model.text}" but was "${text}."`);
```

Comment:
consider:
new Error(`Mismatch: ${JSON.stringify({ expected, actual })}`)
easier to read/diff, and more tolerant of longer text.